libnavui-util/src/main/java/com/mapbox/navigation/ui/utils/internal/resource/DefaultResourceLoader.kt

```diff
@@ -43,20 +48,33 @@ internal class DefaultResourceLoader(
     ): Long {
         val requestId = nextRequestId.incrementAndGet()
         val callbackAdapter = CallbackAdapter(request, callback, observers)
+        val requiresNetwork = request.networkRestriction != NetworkRestriction.DISALLOW_ALL
 
         callbackAdapter.notifyOnStart(request)
-        cancelableMap[requestId] = tileStore.loadResource(
-            /* description */ request.toResourceDescription(),
-            /* options */ request.toResourceLoadOptions("DefaultResourceLoader-$requestId"),
-            /* progressCallback */ callbackAdapter
-        ) {
-            cancelableMap.remove(requestId)
-            callbackAdapter.run(it)
+        // Since the TileStore (commonSDK 21.3.1) will defer any requests that require internet connection,
+        // we must verify network requirement here and fail fast if the network is not available.
+        if (requiresNetwork && !reachability.isReachable) {
+            callbackAdapter.run(createError(connectionError()))
```

Comment:
This is too board of a check. If we're offline, we still want to check the cache and see whether the resource is available and only fail if it isn't (so that we can fallback to TTS). It looks like we'll need an additional change from Common SDK in form of a flag to fail resource loading if cache is empty and there's no internet connection.

I wouldn't say it's a blocker for the PR as a whole but it would have to be fixed before we call the feature complete.

Reply:
When we are offline, we could try and load the cached resource by calling the `TileStore` with `ResourceLoadFlags.ACCEPT_EXPIRED` and `NetworkRestriction.DISALLOW_ALL` flags. The `TileStore` would then return only the available cached resource without accessing the network.

Reply:
Good idea! I would still hope that an SDK wouldn't have to check the connectivity status when dealing with a resource provider, but this approach sounds like a nice stop-gap. Let's do that and at the same time cut a Common SDK feature request to track for the future.

Reply:
Unfortunately, the  [suggested fix above](https://github.com/mapbox/mapbox-navigation-android/pull/5790#discussion_r875907681) doesn't work. 

For some reason, when in offline mode, calling the `TilesStore::loadResource` with `ResourceLoadFlags.ACCEPT_EXPIRED` and `NetworkRestriction.DISALLOW_ALL` flags doesn't tigger callback with `Unsatisfied` error  as expected (according to [source](https://github.com/mapbox/mapbox-sdk-common/blob/v21.3.1/components/tile_store/src/mapbox/common/tile_store_worker.cpp#L700)).

Instead, the `TileStore` logs an error to the console.

```
W/Mapbox: [tile_store] id=1344, tag=DefaultResourceLoader-6, resource=navigation:https://api.mapbox.com/voice/v1/speak/%3Cspeak%3E%3Camazon:effect%20name=%22drc%22%3E%3Cprosody%20rate=%221.08%22%3EIn%20300%20meters,%20Turn%20right%20onto%20%3Csay-as%20interpret-as=%22address%22%3ELake%20Shore%20Boulevard%20West%3C%2Fsay-as%3E.%3C%2Fprosody%3E%3C%2Famazon:effect%3E%3C%2Fspeak%3E?textType=ssml&language=en&access_token=REDACTED&sku=10apWNHMx4JS4: failed with [NetworkError] Error happened during okhttp download session: Not allowed: [ConnectionError] Not allowed; delaying for 1s after 1 attempts; 1 downloads waiting
```

I think we should merge this PR as is and cut the follow-up ticket for updating the DefaultResourceLoader implementation later once Common SDK has been updated.

Reply:
Logged this issue in CommonSDK repo https://github.com/mapbox/mapbox-sdk-common/issues/2744

Reply:
Now that this suggested stop-gap solution does work, 
> Good idea! I would still hope that an SDK wouldn't have to check the connectivity status when dealing with a resource provider, but this approach sounds like a nice stop-gap.Let's do that and at the same time cut a Common SDK feature request to track for the future.

This still holds true. Did we already create an improvement ticket for this on Common SDK side? @tomaszrybakiewicz